File: front/src/lib/schemas/datasets.ts
"use client";

import { z } from "zod";

import {
  LicenseSchema,
  ProjectSchema,
  RecorderSchema,
  SiteSchema,
} from "./metadata";
import { FileSchema } from "./common";

export const VisibilityLevelSchema = z.enum(["public", "restricted"]);
export type VisibilityLevel = z.infer<typeof VisibilityLevelSchema>;

const optionalTrimmedString = z
  .string()
  .optional()
  .transform((value) => {
    if (value === undefined) {
      return undefined;
    }
    const trimmed = value.trim();
    return trimmed === "" ? undefined : trimmed;
  });

const optionalNullableTrimmedString = z
  .string()
  .nullable()
  .optional()
  .transform((value) => {
    if (value === undefined) {
      return undefined;
    }
    if (value === null) {
      return null;
    }
    const trimmed = value.trim();
    return trimmed === "" ? null : trimmed;
  });

export const DatasetSchema = z.object({
  uuid: z.string().uuid(),
  id: z.number().int().optional(),
  name: z.string(),
  audio_dir: z.string(),
  description: z.string().nullable().optional(),
  recording_count: z.number().int().default(0),
  recording_start_date: z.coerce.date().nullable().optional(),
  recording_end_date: z.coerce.date().nullable().optional(),
  created_on: z.coerce.date(),
  visibility: VisibilityLevelSchema,
  created_by_id: z.string().uuid(),
  project_id: z.string(),
  primary_site_id: z.string().nullable().optional(),
  primary_recorder_id: z.string().nullable().optional(),
  license_id: z.string().nullable().optional(),
  doi: z.string().nullable().optional(),
  note: z.string().nullable().optional(),
  gain: z.number().nullable().optional(),
  project: ProjectSchema.nullable().optional(),
  primary_site: SiteSchema.nullable().optional(),
  primary_recorder: RecorderSchema.nullable().optional(),
  license: LicenseSchema.nullable().optional(),
});

export const DatasetCreateSchema = z.object({
  uuid: z.string().uuid().optional(),
  name: z.string().min(1),
  audio_dir: z.string().min(1, "Select an audio directory"),
  description: optionalTrimmedString,
  visibility: VisibilityLevelSchema.default("restricted"),
  project_id: z.string().min(1, "Select a project"),
  primary_site_id: optionalNullableTrimmedString,
  primary_recorder_id: optionalNullableTrimmedString,
  license_id: optionalNullableTrimmedString,
  doi: optionalTrimmedString,
  note: optionalTrimmedString,
  gain: z.number().nullable().optional(),
});

export const DatasetUpdateSchema = z.object({
  name: z.string().min(1).optional(),
  description: optionalTrimmedString,
  visibility: VisibilityLevelSchema.optional(),
  project_id: optionalNullableTrimmedString,
  primary_site_id: optionalNullableTrimmedString,
  primary_recorder_id: optionalNullableTrimmedString,
  license_id: optionalNullableTrimmedString,
  doi: optionalNullableTrimmedString,
  note: optionalTrimmedString,
  gain: z.number().nullable().optional(),
});

export const DatasetImportSchema = z.object({
  dataset: FileSchema,
  audio_dir: z.string(),
});

export const DatasetCandidateSchema = z.object({
  name: z.string(),
  relative_path: z.string(),
  absolute_path: z.string(),
});

export const DatasetCandidateInfoSchema = z.object({
  relative_path: z.string(),
  absolute_path: z.string(),
  has_nested_directories: z.boolean(),
  audio_file_count: z.number().int(),
});

export const DatasetRecordingSiteSchema = z.object({
  h3_index: z.string().nullable().optional(),
  latitude: z.number(),
  longitude: z.number(),
  recording_count: z.number().int(),
  label: z.string().nullable().optional(),
});

export const DatasetRecordingCalendarBucketSchema = z.object({
  date: z.coerce.date(),
  count: z.number().int(),
});

export const DatasetRecordingHeatmapCellSchema = z.object({
  date: z.coerce.date(),
  hour: z.number().int(),
  count: z.number().int(),
  duration_minutes: z.number(),
});

export const DatasetRecordingTimelineSegmentSchema = z.object({
  recording_uuid: z.string().uuid(),
  start: z.coerce.date(),
  end: z.coerce.date(),
  path: z.string(),
});

export const DatasetOverviewStatsSchema = z.object({
  recording_sites: z.array(DatasetRecordingSiteSchema),
  recording_calendar: z.array(DatasetRecordingCalendarBucketSchema),
  recording_heatmap: z.array(DatasetRecordingHeatmapCellSchema),
  recording_timeline: z.array(DatasetRecordingTimelineSegmentSchema),
  total_duration_seconds: z.number().nullable().optional(),
});

export const DatasetDatetimePatternSchema = z.object({
  dataset_id: z.number().int(),
  pattern_type: z.enum(["strptime", "regex"]),
  pattern: z.string().min(1),
  sample_filename: z.string().nullable().optional(),
  sample_result: z.coerce.date().nullable().optional(),
  uuid: z.string().uuid().optional(),
  created_on: z.coerce.date().optional(),
});

export const DatasetDatetimePatternUpdateSchema = z.object({
  pattern_type: z.enum(["strptime", "regex"]),
  pattern: z.string().min(1),
  sample_filename: z.string().optional(),
});

File: back/src/echoroo/schemas/datasets.py
"""Schemas for handling Datasets."""

import datetime
from enum import Enum
from pathlib import Path
from uuid import UUID

from pydantic import BaseModel, DirectoryPath, Field, model_validator

from echoroo.schemas.base import BaseSchema
from echoroo.schemas.metadata import License, Project, Recorder, Site
from echoroo.schemas.recordings import Recording
from echoroo.models.dataset import VisibilityLevel
from echoroo.models.datetime_pattern import DatetimePatternType

__all__ = [
    "VisibilityLevel",
    "Dataset",
    "DatasetRecording",
    "DatasetCreate",
    "DatasetUpdate",
    "DatasetRecordingCreate",
    "FileState",
    "DatasetCandidate",
    "DatasetCandidateInfo",
    "DatasetDatetimePattern",
    "DatasetDatetimePatternUpdate",
    "DatasetRecordingSite",
    "DatasetRecordingCalendarBucket",
    "DatasetRecordingHeatmapCell",
    "DatasetRecordingTimelineSegment",
    "DatasetOverviewStats",
]


class DatasetCreate(BaseModel):
    """Schema for Dataset objects created by the user."""

    audio_dir: DirectoryPath
    """The path to the directory containing the audio files."""

    name: str = Field(..., min_length=1)
    """The name of the dataset."""

    description: str | None = Field(None)
    """The description of the dataset."""

    visibility: VisibilityLevel = VisibilityLevel.RESTRICTED
    """Visibility level for the dataset."""

    project_id: str = Field(...)
    """Project identifier associated with the dataset."""

    primary_site_id: str | None = Field(default=None)
    """Primary site identifier."""

    primary_recorder_id: str | None = Field(default=None)
    """Primary recorder identifier."""

    license_id: str | None = Field(default=None)
    """License identifier."""

    doi: str | None = Field(default=None)
    """Optional DOI reference."""

    note: str | None = Field(default=None)
    """Optional free-form note."""

    gain: float | None = Field(default=None)
    """Recorder gain in dB."""

    @model_validator(mode="after")
    def _ensure_project_binding(self):
        if not self.project_id:
            raise ValueError("project_id is required for dataset creation")
        return self


class Dataset(BaseSchema):
    """Schema for Dataset objects returned to the user."""

    uuid: UUID
    """The uuid of the dataset."""

    id: int
    """The database id of the dataset."""

    audio_dir: Path
    """The path to the directory containing the audio files."""

    name: str
    """The name of the dataset."""

    description: str | None
    """The description of the dataset."""

    recording_count: int = 0
    """The number of recordings in the dataset."""

    recording_start_date: datetime.datetime | None = None
    """Earliest recording datetime in the dataset."""

    recording_end_date: datetime.datetime | None = None
    """Latest recording datetime in the dataset."""

    visibility: VisibilityLevel
    """Visibility level for the dataset."""

    created_by_id: UUID
    """User who created the dataset."""

    project_id: str
    """Project identifier associated with the dataset."""

    primary_site_id: str | None = None
    """Primary site identifier."""

    primary_recorder_id: str | None = None
    """Primary recorder identifier."""

    license_id: str | None = None
    """License identifier."""

    doi: str | None = None
    """Optional DOI reference."""

    note: str | None = None
    """Optional free-form note."""

    gain: float | None = None
    """Recorder gain in dB."""

    project: Project | None = None
    """Hydrated project metadata."""

    primary_site: Site | None = None
    """Hydrated site metadata."""

    primary_recorder: Recorder | None = None
    """Hydrated recorder metadata."""

    license: License | None = None
    """Hydrated license metadata."""


class DatasetUpdate(BaseModel):
    """Schema for Dataset objects updated by the user."""

    audio_dir: DirectoryPath | None = None
    """The path to the directory containing the audio files."""

    name: str | None = Field(default=None, min_length=1)
    """The name of the dataset."""

    description: str | None = None
    """The description of the dataset."""

    visibility: VisibilityLevel | None = None
    """Updated visibility level."""

    project_id: str | None = None
    """Project identifier associated with the dataset."""

    primary_site_id: str | None = None
    """Primary site identifier."""

    primary_recorder_id: str | None = None
    """Primary recorder identifier."""

    license_id: str | None = None
    """License identifier."""

    doi: str | None = None
    """Optional DOI reference."""

    note: str | None = None
    """Optional free-form note."""

    gain: float | None = None
    """Recorder gain in dB."""


class FileState(Enum):
    """The state of a file in a dataset.

    Datasets can contain files that are not registered in the database. This
    can happen if the file was added to the dataset directory after the
    dataset was registered. Additionally, files can be registered in the
    database but missing from the dataset directory. This can happen if the
    file was removed from the dataset directory after the dataset was
    registered.

    The state of a file can be one of the following:

    - ``missing``: The file is not registered in the database and is missing.

    - ``registered``: The file is registered in the database and is present.

    - ``unregistered``: The file is not registered in the database but is
        present in the dataset directory.
    """

    MISSING = "missing"
    """If the recording is registered but the file is missing."""

    REGISTERED = "registered"
    """If the recording is registered and the file is present."""

    UNREGISTERED = "unregistered"
    """If the recording is not registered but the file is present."""


class DatasetFile(BaseModel):
    """Schema for DatasetFile objects returned to the user."""

    path: Path
    """The path to the file."""

    state: FileState
    """The state of the file."""


class DatasetRecordingCreate(BaseModel):
    """Schema for DatasetRecording objects created by the user."""

    path: Path
    """The path to the recording in the dataset directory."""


class DatasetRecording(BaseSchema):
    """Schema for DatasetRecording objects returned to the user."""

    recording: Recording
    """The uuid of the recording."""

    state: FileState = Field(default=FileState.REGISTERED)
    """The state of the file."""

    path: Path
    """The path to the recording in the dataset directory."""


class DatasetCandidate(BaseModel):
    """Schema for dataset directory candidates detected on disk."""

    name: str
    """Human-friendly label for the directory (folder name)."""

    relative_path: Path
    """Path relative to the configured audio directory."""

    absolute_path: Path
    """Absolute path to the directory on disk."""


class DatasetCandidateInfo(BaseModel):
    """Additional information about a dataset directory candidate."""

    relative_path: Path
    """Path relative to the configured audio directory."""

    absolute_path: Path
    """Absolute path to the directory on disk."""

    has_nested_directories: bool = False
    """True when the directory contains at least one subdirectory."""

    audio_file_count: int = 0
    """Number of audio files detected (wav/mp3/flac and similar)."""


class DatasetRecordingSite(BaseModel):
    """Aggregate information about recording sites for overview maps."""

    h3_index: str | None = None
    """H3 cell identifier if recordings inherit a site hex."""

    latitude: float
    """Latitude of the site centroid."""

    longitude: float
    """Longitude of the site centroid."""

    recording_count: int = 0
    """Number of recordings captured at this site."""

    label: str | None = None
    """Optional human friendly label (e.g. site name)."""


class DatasetRecordingCalendarBucket(BaseModel):
    """Daily recording counts for calendar visualizations."""

    date: datetime.date
    """Day represented by the bucket."""

    count: int = 0
    """Number of recordings captured on that day."""


class DatasetRecordingHeatmapCell(BaseModel):
    """Single cell in the date-time heatmap grid."""

    date: datetime.date
    """Day of the recording."""

    hour: int
    """Hour of day (0-23)."""

    count: int = 0
    """Number of recordings in this date-hour slot."""

    duration_minutes: float = 0.0
    """Total duration of recordings in minutes."""


class DatasetRecordingTimelineSegment(BaseModel):
    """A single recording segment for timeline visualization."""

    recording_uuid: str
    """UUID of the recording."""

    start: datetime.datetime
    """Start datetime of the recording."""

    end: datetime.datetime
    """End datetime of the recording (start + duration)."""

    path: str
    """File path of the recording."""


class DatasetOverviewStats(BaseModel):
    """Roll-up statistics for the dataset overview."""

    recording_sites: list[DatasetRecordingSite] = Field(default_factory=list)
    """Recording site aggregates for the overview map."""

    recording_calendar: list[DatasetRecordingCalendarBucket] = Field(default_factory=list)
    """Daily recording counts for the calendar heatmap."""

    recording_heatmap: list[DatasetRecordingHeatmapCell] = Field(default_factory=list)
    """Date-time heatmap cells for PAM schedule visualization."""

    recording_timeline: list[DatasetRecordingTimelineSegment] = Field(default_factory=list)
    """Timeline segments showing exact recording start/end times."""

    total_duration_seconds: float | None = None
    """Total duration of all recordings in seconds."""

    absolute_path: Path
    """Absolute path to the directory on disk."""

    has_nested_directories: bool = False
    """True when the directory contains at least one subdirectory."""

    audio_file_count: int = 0
    """Number of audio files detected (wav/mp3/flac and similar)."""


class DatasetDatetimePattern(BaseSchema):
    """Persisted datetime parsing configuration for a dataset."""

    dataset_id: int
    pattern_type: DatetimePatternType
    pattern: str
    sample_filename: str | None = None
    sample_result: datetime.datetime | None = None


class DatasetDatetimePatternUpdate(BaseModel):
    """Payload to define or update datetime parsing for a dataset."""

    pattern_type: DatetimePatternType = DatetimePatternType.STRPTIME
    pattern: str = Field(..., min_length=1)
    sample_filename: str | None = None
